File: 15-String/java/03-RomanToInteger.java
class Solution {
    public int romanToInt(String s) {
        int[] values = new int[128];
        values['I'] = 1;
        values['V'] = 5;
        values['X'] = 10;
        values['L'] = 50;
        values['C'] = 100;
        values['D'] = 500;
        values['M'] = 1000;

        int result = 0;
        int n = s.length();

        for (int i = 0; i < n; i++) {
            int current = values[s.charAt(i)];

            if (i < n - 1 && current < values[s.charAt(i + 1)]) {
                result -= current;
            } else {
                result += current;
            }
        }

        return result;
    }
}

File: 15-String/cpp/03-roman-to-integer.cpp
class Solution {
public:
    int romanToInt(string s) {
        unordered_map<char, int> values = {
            {'I', 1}, {'V', 5}, {'X', 10}, {'L', 50},
            {'C', 100}, {'D', 500}, {'M', 1000}
        };

        int result = 0;
        int n = s.size();

        for (int i = 0; i < n; i++) {
            int current = values[s[i]];
            if (i < n - 1 && current < values[s[i + 1]]) {
                result -= current;
            } else {
                result += current;
            }
        }

        return result;
    }
};
